### ISSUE TITLE
Alignment issues when using Eigentypes

### ISSUE THREAD

**Issue by MPJansen** [flexible-collision-library/fcl#152]
Hiya,

I have recently started using your awesome library to do more with robotics motion planning, thanks!

From some pull requests I understand that you have recently switched to using Eigen types under the hood. 
When running the fcl_collicion test,  it sometimes runs into a segmentation error which points to the following Eigen related website:
http://eigen.tuxfamily.org/dox-devel/group__TopicUnalignedArrayAssert.html

My guess is, that when calling 
`
fcl::generateRandomTransforms(extents, transforms, n);
`

transforms is sometimes not 16-bit aligned allocated. The Eigen library provides a workaround in this case:
http://eigen.tuxfamily.org/dox-devel/group__TopicStlContainers.html

I haven't used the library enough to know if this is a problem outside of this test case, but I thought that since the switch to eigen is fairly recent, it might be worth knowing.

Kind regards!

**Comment by sherm1** [comment]
@MPJansen, what platform were you running on when you saw that seg fault, and was it 32 or 64 bits?

**Comment by MPJansen** [comment]
@sherm1 ,

I am running the examples from a windows 10 computer under vs 2015, 32 bit debug comp.
The exact settings are on my work computer which I do not have access to right this sec.

I mistakenly called it a segmentation fault though, it is actually an Eigen debug assert that tests for alignment issues, like the one detailed at
http://eigen.tuxfamily.org/dox-devel/group__TopicStlContainers.html

A similar thing happens when you use fixed size (stack allocated) matrices as struct or new'ed class members without special care. The wrapping inside such a container messes up the memory alignment which in turn will prohibit simd instructions on these matrices. 

Luckily the Eigen library provides methods to circumvent these situations.

Hope this helps,

**Comment by chhtz** [comment]
This is a known issue with Eigen in `std` containers. The two possible fixes are to use the patched `std::vector` as suggested in the link, or to generally use unaligned Eigen types (since you `typedef` them anyway, you could easily write e.g.,
```
template <typename Scalar> using Isometry3 =
  Eigen::Transform<Scalar, 3, Eigen::Isometry | Eigen::DontAlign>;
```

The reason this does not occur on 64-bit systems is that they by default align everything to 16 bytes (you will again have problems when compiling with the most recent Eigen version and AVX enabled)

**Comment by MPJansen** [comment]
From scanning the source files a little bit, some of the FCL internals seem to be using std containers in conjunction with eigen fixed size types, i.e: 

`generateBVHModel(BVHModel<BV>& model, const Sphere& shape, const Transform3d& pose, unsigned int seg, unsigned int ring)`

has its points defined as
`  std::vector<Vector3d> points; `

Which I guess should be
`#include <Eigen/StdVector>`
`std::vector<Eigen::Vector3d,Eigen::aligned_allocator<Eigen::Vector3d> > points`

Disabling alignment would also disable vectorization of calculations using these vectors (unless a local aligned copy is made) 
Maybe @jslee02 would be interested?

Kind regards,

**Comment by jslee02** [comment]
Oh, I forgot to mention here. I'm trying to resolving this issue in #154 together.

**Comment by chhtz** [comment]
FYI: With C++11 the `#include <Eigen/StdVector>` is not necessary anymore. Also, `Vector3d` will not be aligned anyway, so for this it will not make a difference. And, with the very newest Eigen version there will be some vectorization even for unaligned types (because newer CPUs have barely overhead for unaligned loads) -- this can be deactivated, if wanted (for older CPUs).

**Comment by OwlCoder** [comment]
@jslee02 I guess Eigen::Isometry (mentioned as a parameter in the template for Isometry3) is deprecated. I can't find any information on the internet. Are you still working on it? Or is it already fixed and there's some other issue?

**Comment by jslee02** [comment]
I believe the alignment issue is addressed in #154.

**Comment by MPJansen** [comment]
Looks great!
